Sources/ConfidenceProvider/Http/NetworkClient.swift

```diff
@@ -44,67 +44,56 @@ final class NetworkClient: HttpClient {
     func post<T: Decodable>(
         path: String,
         data: Codable,
-        completion: @escaping (HttpClientResult<T>) -> Void
-    ) throws {
+        completion: @escaping (HttpClientResult<T>) async -> Void
+    ) async throws {
         let request = try buildRequest(path: path, data: data)
-        perform(request: request, retry: self.retry) { response, data, error in
+        await perform(request: request, retry: self.retry) { response, data, error in
             if let error {
-                completion(.failure(error))
+                await completion(.failure(error))
                 return
             }
 
             guard let response, let data else {
-                completion(.failure(ConfidenceError.internalError(message: "Bad response")))
+                await completion(.failure(ConfidenceError.internalError(message: "Bad response")))
                 return
             }
 
             do {
                 let httpClientResult: HttpClientResponse<T> =
-                    try self.buildResponse(response: response, data: data)
-                completion(.success(httpClientResult))
+                try self.buildResponse(response: response, data: data)
+                await completion(.success(httpClientResult))
             } catch {
-                completion(.failure(error))
+                await completion(.failure(error))
             }
         }
     }
 
     private func perform(
         request: URLRequest,
         retry: Retry,
-        completion: @escaping (HTTPURLResponse?, Data?, Error?) -> Void
-    ) {
+        completion: @escaping (HTTPURLResponse?, Data?, Error?) async -> Void
+    ) async {
         let retryHandler = retry.handler()
         let retryWait: TimeInterval? = retryHandler.retryIn()
 
-        self.session.dataTask(with: request) { data, response, error in
-            if let error {
-                if self.shouldRetry(error: error), let retryWait {
-                    DispatchQueue.main.asyncAfter(deadline: .now() + retryWait) {
-                        self.perform(request: request, retry: retry, completion: completion)
-                    }
-                    return
-                } else {
-                    completion(nil, nil, error)
-                }
-            }
-
+        do {
+            let (data, response) = try await self.session.data(for: request)
```

Comment:
`error` is not returned from `.data()`, while that was the case with `.dataTask()`. However, `.data()` is marked as throwing, so the do/catch wrapping _should_ in practice replicate the exact same behaviour of `.dataTask()`